File: engine/src/main/java/de/hanno/hpengine/util/commandqueue/CommandQueue.java
package de.hanno.hpengine.util.commandqueue;

import java.util.Iterator;
import java.util.concurrent.*;
import java.util.logging.Logger;

import static de.hanno.hpengine.engine.threads.UpdateThread.isUpdateThread;

public class CommandQueue {
    private static final Logger LOGGER = Logger.getLogger(CommandQueue.class.getName());

    private ConcurrentLinkedQueue<FutureCallable> workQueue = new ConcurrentLinkedQueue<>();

    public boolean executeCommands() {
        boolean executedCommands = false;
        while(executeCommand()) {
            executedCommands = true;
        }
        return executedCommands;
    }

    public boolean executeCommand() {
        FutureCallable command = workQueue.poll();
        if(command != null) {
            try {
                command.complete(command.execute());
                return true;
            } catch (Exception e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return false;
        }
    }

    public <RESULT_TYPE> CompletableFuture<RESULT_TYPE> addCommand(Runnable runnable) {
        FutureCallable command = new RunnableCallable(runnable);
        return addCommand(command);
    }
    public <RESULT_TYPE> CompletableFuture<RESULT_TYPE> addCommand(FutureCallable<RESULT_TYPE> command) {
        if(executeDirectly()) {
            try {
                command.getFuture().complete(command.execute());
                return command.getFuture();
            } catch (Exception e) {
                e.printStackTrace();
                command.getFuture().completeExceptionally(e);
                return command.getFuture();
            }
        }
        workQueue.offer(command);
        return command.getFuture();
    }
    public Exception execute(Runnable runnable, boolean andBlock) {
        if(executeDirectly()) {
            runnable.run();
            return null;
        }
        CompletableFuture<Object> future = addCommand(new RunnableCallable(runnable));

        if(andBlock) {
            future.join();
        }
        return null;
    }
    public int size() {
        return workQueue.size();
    }

    public Iterator<FutureCallable> getIterator() {
        return workQueue.iterator();
    }

    public ConcurrentLinkedQueue<FutureCallable> getWorkQueue() {
        return workQueue;
    }

    private static class RunnableCallable extends FutureCallable<Object> {

        private final Runnable runnable;

        public RunnableCallable(Runnable runnable) {
            this.runnable = runnable;
        }

        @Override
        public Object execute() throws Exception {
            runnable.run();
            return null;
        }
    }

    protected boolean executeDirectly() { return false; }
}

File: engine/src/main/java/de/hanno/hpengine/util/commandqueue/CommandQueue.kt
package de.hanno.hpengine.util.commandqueue

import kotlinx.coroutines.*
import java.util.concurrent.Callable
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.logging.Level
import java.util.logging.Logger

open class CommandQueue @JvmOverloads constructor(executorService: ExecutorService, private val executeDirectly: () -> Boolean = { false }) {
    val executor = executorService
    val dispatcher = executor.asCoroutineDispatcher()

    fun executeCommands(): Boolean {
        runBlocking {
            launch(dispatcher) {
                try {
                    yield()
                } catch (e: Error) {
                    LOGGER.log(Level.SEVERE, "", e)
                }
            }
        }
        return true
    }

    fun addCommand(runnable: Runnable): CompletableFuture<Void> {
        val result = CompletableFuture<Void>()
        execute(Runnable { runnable.run().apply { result.complete(null) } } , true)
        return result
    }

    fun <RESULT_TYPE> addCommand(command: FutureCallable<RESULT_TYPE>): CompletableFuture<RESULT_TYPE> {
        if (executeDirectly.invoke()) {
            try {
                command.future.complete(command.execute())
            } catch (e: Exception) {
                e.printStackTrace()
                command.future.completeExceptionally(e)
            }
        } else {
            execute(Runnable { command.execute().also { command.future.complete(it) } }, false)
        }

        return command.future
    }

    fun <RESULT_TYPE> calculate(callable: Callable<RESULT_TYPE>): RESULT_TYPE? {
        if(executeDirectly()) {
            return callable.call()
        }
        return runBlocking {
            withContext(dispatcher) {
                callable.call()
            }
        }
    }

    fun execute(runnable: Runnable, andBlock: Boolean) {
        if (executeDirectly.invoke()) {
            runnable.run()
        }
        if(andBlock) {
            runBlocking {
                withContext(dispatcher) {
                    runnable.run()
                }
            }
        } else {
            GlobalScope.launch(dispatcher) {
                runnable.run()
            }
        }
    }

    companion object {
        private val LOGGER = Logger.getLogger(CommandQueue::class.java.name)
    }
}
